src/future/builtin-fields/AutoFields.js

```diff
@@ -1,17 +1,22 @@
-import React, { useContext } from 'react';
-import { RenderContext } from '@/src/future/context';
+import React from 'react';
+import { useReactiveValueMeta } from '@/src/future/ReactiveValue';
 import { startCase } from '@/src/stringUtils';
 
-const inferSchema = ([key, value]) => ({
+const inferSchema = ({ key, type }) => ({
   key,
   label: startCase(key),
-  type: typeof value,
+  type,
 });
 
 export default function createAutoFields(defaultFieldComponents) {
   return function AutoFields(props) {
-    const { initialValue: formValues } = useContext(RenderContext);
-    const schema = Object.entries(formValues).map(inferSchema);
+    const meta = useReactiveValueMeta();
```

Comment:
@stevelikesmusic I made an [even more performant `ReactiveValue`](https://github.com/zapier/formatic/pull/124/commits/65f7c083219f7e37e19fae671bc21ac77ddfc0b1). (Unfortunately, it did get a _bit_ more complex, but it's still nicely isolated behind hooks.) And as part of that, I added a `useReactiveValueMeta` that I swapped out here. It gives you access to property types and only updates when those types change. This is also dynamic, so new properties that are added will generate new auto fields.

I _think_ this may be roughly how validation errors will work once we get to that. Effectively, validation errors will just be extra metadata this sits alongside the values.